### PR TITLE
Add NEON intrinsics

### PR DESCRIPTION
**Pull request by zzlei**
Please **DON'T** merge this pull request. I'm bringing here for some discussions.

### CODE REVIEW

File: src/pseudo_intrinsics.h [hunk 1]
```diff
@@ -490,35 +555,14 @@ typedef __m64i vtype;
 #endif /* __SSE3__ || __MIC__ */
```
Comment (zzlei) [inline src/pseudo_intrinsics.h]: Does this kind of optimization really make sense?

File: src/pseudo_intrinsics.h [hunk 2]
```diff
@@ -34,8 +34,57 @@
...
+#ifdef __ARM_NEON__
+#include <arm_neon.h>
+
+typedef uint32x4_t vtype32;
+typedef uint64x2_t vtype64;
+typedef vtype64 vtype;
+
+#define vadd_epi32              vaddq_u32
+#define vadd_epi64              vaddq_u64
+#define vand                    vandq_u32
```
Comment (zzlei) [inline src/pseudo_intrinsics.h]: This is troublesome, just like in AltiVec. Although `vandq_u32` and `vandq_u64` translate to the same instruction, they have different function signatures, thus are not interchangeable.

I'm thinking maybe we should have different versions of `vand`, `vload`, etc, which would make it more natural to write AltiVec and NEON intrinsics. So far only x86 uses the type-agnostic form of those intrinsics. 

We could also distinguish between `vtype32` and `vtype64`, which may improve readability (knowing what exactly type of data we are dealing with).

**Comment by zzlei** [comment]
I refactored pseudo_intrinsics.h a bit. Those `#ifdef`s for CPU dispatching make me kind of dizzy...

I think there's still space for improvement of readability. I'll try to refactor it further after NEON support is completed.

File: src/pseudo_intrinsics.h (hunk 1 above)
Comment (magnumripper) [inline src/pseudo_intrinsics.h]: You mean the one "for really old CPUs"? I'm not sure of its origin but I guess it makes sense. Anything that can save a single CPU cycle is...

Comment (magnumripper) [inline src/pseudo_intrinsics.h]: BTW please not that the ternary function is supposed to be optimized away completely, so the code will be just either of the functions in each place this macro is used. Maybe you didn't realize that.

Comment (zzlei) [inline src/pseudo_intrinsics.h]: So the argument `s` is always a constant in the code. That makes sense.

**Pull request closed without merging**